From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Igor Sheludko <isheludko@gmail.com>
Date: Tue, 23 Jun 2026 09:33:35 +0200
Subject: [v8] Prepare for migrating v8::MicrotaskQueue to cppgc

Backports https://github.com/v8/node/commit/21c0130e8a52fe3c42d09372cd679a78bd7dc5eb
Refs https://chromium-review.googlesource.com/c/v8/v8/+/7991718

v8::MicrotaskQueue is now always a CppGC-managed object and the
V8_CPPGC_MICROTASK_QUEUE define no longer exists, so the CppGC form is
used unconditionally.
Refs https://chromium-review.googlesource.com/c/v8/v8/+/8336255

Will be removed when Node.js bumps to version of V8 that contains
the API change.

diff --git a/src/node_contextify.cc b/src/node_contextify.cc
index b03b19f4b6f84332766a3e1192235e0c53dbfcc1..40d1cdc169603e7617639eb38b7767677aca08ce 100644
--- a/src/node_contextify.cc
+++ b/src/node_contextify.cc
@@ -127,7 +127,7 @@ ContextifyContext* ContextifyContext::New(Environment* env,
 
   MicrotaskQueue* queue =
       options->own_microtask_queue
-          ? options->own_microtask_queue.get()
+          ? options->own_microtask_queue
           : env->isolate()->GetCurrentContext()->GetMicrotaskQueue();
 
   Local<Context> v8_context;
@@ -148,9 +148,7 @@ ContextifyContext::ContextifyContext(Environment* env,
                                      Local<Object> wrapper,
                                      Local<Context> v8_context,
                                      ContextOptions* options)
-    : microtask_queue_(options->own_microtask_queue
-                           ? options->own_microtask_queue.release()
-                           : nullptr) {
+    : microtask_queue_(options->own_microtask_queue) {
   CppgcMixin::Wrap(this, env, wrapper);
 
   context_.Reset(env->isolate(), v8_context);
diff --git a/src/node_contextify.h b/src/node_contextify.h
index 1066633daf65d430b5f46e1b226ee3cf70979c33..c7c39ae8c0632695d8b22023b8e6b1d1cb2e6589 100644
--- a/src/node_contextify.h
+++ b/src/node_contextify.h
@@ -18,7 +18,7 @@ struct ContextOptions {
   v8::Local<v8::String> origin;
   v8::Local<v8::Boolean> allow_code_gen_strings;
   v8::Local<v8::Boolean> allow_code_gen_wasm;
-  std::unique_ptr<v8::MicrotaskQueue> own_microtask_queue;
+  v8::MicrotaskQueue* own_microtask_queue = nullptr;
   v8::Local<v8::Symbol> host_defined_options_id;
   bool vanilla = false;
 };
@@ -122,7 +122,7 @@ class ContextifyContext final : CPPGC_MIXIN(ContextifyContext) {
   }
 
   inline v8::MicrotaskQueue* microtask_queue() const {
-    return microtask_queue_.get();
+    return microtask_queue_;
   }
 
   template <typename T>
@@ -186,7 +186,7 @@ class ContextifyContext final : CPPGC_MIXIN(ContextifyContext) {
       const v8::PropertyCallbackInfo<v8::Array>& args);
 
   v8::TracedReference<v8::Context> context_;
-  std::unique_ptr<v8::MicrotaskQueue> microtask_queue_;
+  cppgc::Persistent<v8::MicrotaskQueue> microtask_queue_;
 };
 
 class ContextifyScript final : CPPGC_MIXIN(ContextifyScript) {
diff --git a/test/cctest/test_environment.cc b/test/cctest/test_environment.cc
index a219d512570136a9b1e940942b8426bf19d0d1f1..3e846a672900807ba8df19e626837ce18be503c4 100644
--- a/test/cctest/test_environment.cc
+++ b/test/cctest/test_environment.cc
@@ -706,15 +706,16 @@ TEST_F(EnvironmentTest, NestedMicrotaskQueue) {
   const v8::HandleScope handle_scope(isolate_);
   const Argv argv;
 
-  std::unique_ptr<v8::MicrotaskQueue> queue = v8::MicrotaskQueue::New(
-      isolate_, v8::MicrotasksPolicy::kExplicit);
+  v8::MicrotaskQueue* queue =
+      v8::MicrotaskQueue::New(isolate_, v8::MicrotasksPolicy::kExplicit);
+
   v8::Local<v8::Context> context =
       v8::Context::New(isolate_,
                        nullptr,
                        {},
                        {},
                        v8::DeserializeInternalFieldsCallback(),
-                       queue.get());
+                       queue);
   node::InitializeContext(context);
   v8::Context::Scope context_scope(context);
 
